doc/1.4/language.md

```diff
@@ -3815,6 +3828,271 @@ independent method callback(int i, void *aux) {
+        ... // An extremely complicated implementation
+    }
+}
+
+template gated_register is register {
+    method write_allowed() -> (bool) default {
+        return true;
+    }
+
+    method on_write_attempted_when_not_allowed() default {
+        log spec_viol: "%s was written to when not allowed", qname;
+    }
+
+    method write_register(uint64 written, uint64 enabled_bytes,
+                          void *aux) default {
+        if (write_allowed()) {
+            default(written, enabled_bytes, aux);
+        } else {
+            on_write_attempted_when_not_allowed();
+        }
+    }
+}
+
+template very_complex_gated_register is (very_complex_register,
+                                         gated_register) {
+    // No sensible way to combine the two implementations by calling both.
+    // Even if there were, calling both implementations would cause each field
+    // of the register to be written to multiple times, potentially duplicating
+    // side-effects, which is undesirable.
+    // Instead, very_complex_register is chosen as the base implementation
+    // called, and the behaviour of gated_register is replicated around that
+    // call.
+    method write_register(uint64 written, uint64 enabled_bytes,
+                          void *aux) default {
+        if (write_allowed()) {
+            this.templates.very_complex_register.write_register(
+                written, enabled_bytes, aux);
+        } else {
+            on_write_attempted_when_not_allowed();
+        }
+    }
+}
+
+in each (gated_register, very_complex_register) {
+    is very_complex_gated_register;
+}
+```
+
+A template-qualified method implementation call is resolved by using
+the method implementation provided to the object by the named template.
+If no such implementation is provided (whether it be because the template does
+not specify one, or specifies one which is not provided to the object due to its
+definition being eliminated by an [`#if`](#conditional-objects)), then the
+ancestor templates of the named template are recursively searched for the
+highest-ranking (most specific) implementation provided by them. If the ancestor
+templates provide multiple hierarchically unrelated implementations, then the
+choice is ambiguous and the call will be rejected by the compiler. In this case,
+the modeller must refine the template-qualified method implementation call to
+name the ancestor template whose implementation they would like to use.
+
+A template-qualified method implementation call done via [a value of template
+type](#templates-as-types) &mdash; most significantly, done via the (implicit)
+`this` within the body of a `shared` method &mdash; function differently in
+comparison to with compile-time object references. The specified template
+must be an ancestor template of the value's template type, the <tt>object</tt>
+template, or the template type itself; and the specified template **must
```

Comment:
either `s/; and/, and/` or `s/; and /; furthermore, /`

Reply:
yeah i struggled with this. I used `;` instead of `,` because I wanted a clear separation between the two conditions, though maybe that's not needed.

Reply:
Then I suggest `furthermore`.